features/netsocket/nsapi_dns.cpp

```diff
@@ -305,3 +519,337 @@ nsapi_error_t nsapi_dns_query(NetworkStack *stack, const char *host,
     address->set_addr(addr);
     return (nsapi_error_t)((result > 0) ? 0 : result);
 }
+
+nsapi_error_t nsapi_dns_query_async(NetworkStack *stack, const char *host,
+        NetworkStack::hostbyname_cb_t callback, void *data, nsapi_version_t version)
+{
+    nsapi_size_or_error_t result = nsapi_dns_query_multiple_async(stack, host, callback, data, 1, version);
+    return (nsapi_error_t)((result > 0) ? 0 : result);
+}
+
+static nsapi_error_t nsapi_dns_call_default(mbed::Callback<void()> func)
+{
+    events::EventQueue *event_queue = mbed::mbed_event_queue();
+    if (!event_queue) {
+        return NSAPI_ERROR_NO_MEMORY;
+    }
+    if (event_queue->call(func) == 0) {
+        return NSAPI_ERROR_NO_MEMORY;
+    }
+    return NSAPI_ERROR_OK ;
+}
+
+static nsapi_error_t nsapi_dns_call_in_default(int delay, mbed::Callback<void()> func)
+{
+    events::EventQueue *event_queue = mbed::mbed_event_queue();
+    if (!event_queue) {
+        return NSAPI_ERROR_NO_MEMORY;
+    }
+    if (event_queue->call_in(delay, func) == 0) {
+        return NSAPI_ERROR_NO_MEMORY;
+    }
+    return NSAPI_ERROR_OK ;
+}
+
+void nsapi_dns_call_set(nsapi_dns_call_t callback)
+{
+    dns_call = callback;
+}
+
+void nsapi_dns_call_in_set(nsapi_dns_call_in_t callback)
+{
+    dns_call_in = callback;
+}
+
+static nsapi_error_t nsapi_dns_call(NetworkStack *stack, mbed::Callback<void()> func)
+{
+    if (stack->onboardNetworkStack()) {
+        OnboardNetworkStack *onboard_stack = reinterpret_cast<OnboardNetworkStack *>(stack);
+        return onboard_stack->call(func);
```

Comment:
Thinking about it more, I think this is just a bit too upside down. We're poking all these APIs through the network stack because we need them from an API that's called internally?

Why not just pass the necessary things to the more-or-less internal API?

If the dns_query_async API needs an event loop, just make that a necessary argument to the function?

Reply:
The logic here is that we've hit a repeating requirement for network-related stuff to do little tasks. Things like PHY status monitoring in Ethernet drivers. Previously  lwip/Nanostack drivers would use lwip/Nanostack event APIs to schedule those. In the EMAC world that's out, and they've been tending to switch to mbed global `EventQueue`s.

Which isn't bad, but it is possible they end up being the only user of the global event queue. It will be the case in the smallest mini client.

So the thought is that we know any onboard network stack must have some sort of event-like mechanism, so this genericises the "run in stack context" drivers have historically used, and it remains the case that it will never spawn an extra thread. We were thinking of Ethernet drivers as much as DNS.

The need for this special API would be eliminated if we required `OnboardNetworkStack`s to be using `EventQueue` for their processing, but that's not the case for either `LWIP` or `Nanostack` at the moment.

Reply:
But these all seem like things that are called from the network stack, so if the network stack has an event queue couldn't it just be passed to the ethernet drivers at call/init time?

Reply:
Yes, it effectively would be for drivers - we're not doing that yet. But it's not necessarily an mbed OS `EventQueue`.

For DNS API purposes the callback context for DNS completion can be API-user specified if they have special requirements for their own event system, or they can say "don't care", in which case they end up getting called in the network stack's context, or the global `EventQueue` which acts as the default for `NetworkStack`.

(The whole DNS state machine runs in a single context, so that user-specified context setting has to be global, not per query. They can always use a trampoline to bounce per call if necessary).

Certainly simplest thing for DNS async API would be to require an `EventQueue` to run on, and default to the global one. But that would mean an extra thread with LWIP and Nanostack, which is what we're trying to avoid. And we're not facing up to restructuring their respective event/thread model to use `EventQueue`.

Reply:
In that case, do you think we need an abstract EventQueue interface that could be backed by different event queue implementation? Then the DNS/EMAC apis could take in an "EventQueueHandle" that could be backed by the native nanostack event queue.

(Name is a joke)

Reply:
Yes, I was thinking something like that might be in order, and was pondering suggesting it. Maybe if we start unambitiously with just the call/call_in API and we can flesh it out later.

(Only reason I was hesitating suggesting it was possible delivery time impact for this PR)

Reply:
Yeah, it would probably take a while. Also not sure it's entirely possible to combine the virutal class with the templating inside the EventQueue.

Maybe you could just pass a pointer to the call_in function?
```
void emac_function(Callback<int(int ms, Callback<void()> cb)> call_in_cb);
```

(you could also use call_in(0, blah) for the call function without a timeout if you want one less argument)

Reply:
I'll change the API so that network stack provides private method to get the callback and that is passed to DNS resolver. That way there will be no public call/call_in interface. Also EMAC can in future can use same kind of interface (as described above).